src/Http/Http/src/Features/FormFeature.cs

```diff
@@ -181,27 +181,29 @@ private async Task<IFormCollection> InnerReadFormAsync(CancellationToken cancell
                         {
                             var fileSection = new FileMultipartSection(section, contentDisposition);
 
-                            // Enable buffering for the file if not already done for the full body
-                            section.EnableRewind(
-                                _request.HttpContext.Response.RegisterForDispose,
-                                _options.MemoryBufferThreshold, _options.MultipartBodyLengthLimit);
-
-                            // Find the end
-                            await section.Body.DrainAsync(cancellationToken);
-
                             var name = fileSection.Name;
                             var fileName = fileSection.FileName;
 
                             FormFile file;
                             if (section.BaseStreamOffset.HasValue)
                             {
+                                await section.Body.DrainAsync(cancellationToken);
+
                                 // Relative reference to buffered request body
                                 file = new FormFile(_request.Body, section.BaseStreamOffset.GetValueOrDefault(), section.Body.Length, name, fileName);
                             }
                             else
                             {
+                                // Enable buffering for the file if not already done for the full body
+                                var stream = section.EnableRewind(
```

Comment:
This API is going to need some work, the old EnableRewind worked in place but this one requires you to track the stream independently.

Elsewhere we've discussed exposing both a stream and pipe on the section object so the consumer could decide which they wanted to consume. Then you'd have somewhere to store the buffer stream.

Reply:
I will try something